### PR TITLE
Index to grouped data frame using `Dict`s

### PR DESCRIPTION
**Pull request by pdeffebach**
Following discussion in #2265, this is the better strategy. 

This PR so far has the implementation and docs, but no tests.

### CODE REVIEW

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:328]: We should access columns of appropriate index (also make sure this is tested when e.g. columns 5,3, and 7 are grouping columns in the order I have given)
```suggestion
        key[T(_names(gd)[gd.cols[i]])]
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:332]: We should not allow mixing `Symbol`s and strings I think:
```suggestion
function Base.to_index(gd::GroupedDataFrame, key::Union{Dict{Symbol},Dict{<:AbstractString}})
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:326]: As below I think it is better to be stricter just to be sure all works OK.
```suggestion
function _dict_to_tuple(key::Union{Dict{Symbol},Dict{<:AbstractString}} , gd)
    T = keytype(d)
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:333]: The second test was incorrect as commented above
```suggestion
    if length(key) != length(gd.cols)
```
Also I think we do not need it as if the key is not found we will get an error trying to retrieve it in `_dict_to_tuple`. Finally I would move the `length`-check to `_dict_to_tuple` to avoid code duplication in `haskey`.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:526]: As commented above - if we move this code to `_dict_to_tuple` then we will not have code duplication.

**Comment by bkamins** [comment]
Thank you for the fix. The design looks good to me. I have left some comments. Also `function Base.to_index(gd::GroupedDataFrame, idxs::AbstractVector{T}) where {T}` needs to be updated to allow `Dict`.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:334]: this check should go to `_dict_to_tuple`.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:527]: I am unresolving this because this check is incorrect. Also if you move the `if` in `to_index` to `_dict_to_tuple` it will not be needed at all.

**Comment by bkamins** [comment]
As I have commented earlier this https://github.com/JuliaData/DataFrames.jl/pull/2281/files#diff-6349421a054bb7e74b7f27ae86304cbaR342 also should be fixed.

**Comment by pdeffebach** [comment]
Sorry about that. Was fixing them but got distracted. 

Should be ready for another review. I added a `Dict` method where the keys are `String`s by default. Then I added a bunch of tests.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:331]: This is needed as for strings we need to have a type that is able to perform the conversion described below
```suggestion
    T = keytype(key) <: Symbol ? Symbol : String
```
This assumes that a `Dict` with any string keys should be able to handle getting `String` input (@nalimilan - do you think it is safe to assume; maybe some strange custom `AbstractStrings` with non UTF-8 encoding would fail here?).

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:529]: this is now not needed, as it is handled in `_dict_to_tuple`

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:545]: maybe add what types for keys of `Dict` are allowed?

File: test/grouping.jl
Comment (bkamins) [inline test/grouping.jl:1342]: Please add some tests for `Dict` where `keytype` is `AbstractString` and `SubString` (to check my fix above).

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:364]: This means we allow mixing `Dict`s with different signatures, but maybe it is OK? What do you think.
If we allow it please add test that something like: `[Dict(:a=>1), Dict("a"=>2)]` (appropriately adjusted) passes.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:306]: Defining it in this way should give a bit better error messages
```suggestion
const GroupKeyTypes = Union{GroupKey, Tuple, NamedTuple, Dict{Symbol}, Dict{<:AbstractString}}
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:364]: But even if we keep it (i.e. allowing mixing) I would restrict it to `Union{Dict{Symbol}, Dict{<:AbstractString}}`. This will not change how the code works, but will produce a nicer error message `"Invalid index ...`. Now for invalid `Dict` you will get a `MethodError` when trying to call `to_index` and probably this will confuse users.

Comment (pdeffebach) [reply src/groupeddataframe/groupeddataframe.jl]: I think it's better to be strict for now. I have added that code that will throw an error for different key types and added a test.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:585]:
```suggestion
function Base.get(gd::GroupedDataFrame, key::Union{Tuple, NamedTuple, Dict{Symbol}, Dict{<:AbstractString}}, default)
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:364]:
```suggestion
    elseif E1 <: Dict{Symbol}
        Dict{Symbol}
    elseif E1 <: Dict{<:AbstractString}
        Dict{<:AbstractString}
```

File: test/grouping.jl
Comment (bkamins) [inline test/grouping.jl:1377]:
```suggestion
    @test gd[Dict(Test.GenericString("a") => :A, "b" => 1)] ≅  gd[1]
```

**Comment by bkamins** [comment]
I have pushed some small things. If CI passes can you please have a final look at it and confirm all is clear and works as intended? Then I think we can approve it (but I will wait with merging for @nalimilan as he has not commented on this API change).

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:270]: This raises again the question of whether we should use symbols or strings by default... Another place where we had to choose is `eachcol`/`DataFrameColumns`. We should have a general policy about this.

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:326]:
```suggestion
function _dict_to_tuple(key::Union{Dict{Symbol}, Dict{<:AbstractString}}, gd)
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:306]: Maybe this should be `AbstractDict`? Same elsewhere.

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:543]:
```suggestion
order as the `cols` argument to `groupby`). It may also be a `Dict` with
`Symbol` or string keys, in which case the order of the arguments does not matter. 
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:587]:
```suggestion
function Base.get(gd::GroupedDataFrame, key::GroupKeyTypes, default)
```

File: src/groupeddataframe/splitapplycombine.jl
Comment (nalimilan) [inline src/groupeddataframe/splitapplycombine.jl:37]:
```suggestion
be slower than using the equivalent `GroupKey`. Finally, a `Dict` with
column names as keys can be used (order does not matter in this case).
Note that indexing with `GroupKey` is faster than other methods.
```
(And remove the mention of slower above for tuples and named tuples.)

File: test/grouping.jl
Comment (nalimilan) [inline test/grouping.jl:1375]:
```suggestion
    # SubString and AbstractString
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:306]: why not. This probably requires a careful consideration if this does not introduce problems though (probably not).

For sure we can provide a conversion from `GroupKey` to a dictionary only for a concrete type (`Dict`).

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:270]: The general policy is to use `Symbol` in this case (which was settled after I looked at this PR).
I know @pdeffebach voted otherwise :smile:, but we should be consistent.

Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:270]: Ah - as @nalimilan prompted me - we have not settled the decision in https://github.com/JuliaData/DataFrames.jl/issues/2298#issuecomment-646703922. So first we should decide there and then update this PR accordingly.

File: src/groupeddataframe/groupeddataframe.jl
Comment (pdeffebach) [reply src/groupeddataframe/groupeddataframe.jl]: this can be resolved, sorry.

**Comment by pdeffebach** [comment]
I've added `AbstractDict` for this, but am waiting on #2298 for the `String` vs `Symbol` output.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:332]:
```suggestion
    return ntuple(length(gd.cols)) do i 
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:213]:
```suggestion
equivalent `Tuple`, `NamedTuple`, or `AbstractDict`.
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:421]:
```suggestion
Unlike the equivalent `Tuple`, `NamedTuple`, and `AbstractDict`, these keys can be used to index
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:542]:
```suggestion
order as the `cols` argument to `groupby`). It may also be a `AbstractDict` with
```

File: src/groupeddataframe/splitapplycombine.jl
Comment (bkamins) [inline src/groupeddataframe/splitapplycombine.jl:36]:
```suggestion
be slower than using the equivalent `GroupKey`. Finally, a `AbstractDict` with
```

File: src/groupeddataframe/splitapplycombine.jl
Comment (bkamins) [inline src/groupeddataframe/splitapplycombine.jl:37]:
```suggestion
column names as keys can be used (order does not matter in this case
even for ordered dictionaries).
```

**Comment by bkamins** [comment]
I think in #2298 we will end up with `Symbol` as the default. But given your comment there let us wait for more feedback. This PR will be merged after the decision in #2298 is made

**Comment by bkamins** [comment]
Given the decision in https://github.com/JuliaData/DataFrames.jl/issues/2298 I think this PR can be finalized following it if you have time. Thank you!

**Comment by bkamins** [comment]
If you feel it would be good to have it can you tell me if you have time to finalize this PR so that we can include it in the coming release?

**Comment by pdeffebach** [comment]
yes, sorry! I have time I will work on it today.

**Comment by pdeffebach** [comment]
I think I may have borked this with a force push that I thought was working well after a `git reset --soft`. 

I will need a bit more time to write tests and make sure the diff is correct and incorporates all previous comments.

**Comment by bkamins** [comment]
OK - thank you for working on it. This is indeed an issue when a PR is stalled for some time. I will review it carefully when you let me know it is ready.

**Comment by pdeffebach** [comment]
This is ready for a review. 

I somehow messed up a lot of this via git. I think I have the commit ready to go now, however, and incorporated the entire previous discussion, whether or not there is a valid git history tying it all together. 

Apologies for this mistake. I know it makes a bit more work for you both.

**Comment by bkamins** [comment]
CI fails on Julia 1.0.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:398]: I would assume this should be enough (If I am right then capturing `T` above is not needed)
```suggestion
        key[string(gd.cols[i])]
```
also why do you prefer `do` syntax here instead of just `i -> key[string(gd.cols[i])]`?

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:395]: Also check if `key` does not contain other entries than these in `gd.cols`. It is enough to test length, as other mismatches will be captured by indexing.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:405]: here using `T` is not needed for sure, as `gd.cols` contains `Symbols`. Other comments as above

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:395]: ah - you check it below, but I think it would be better to check it in these methods and drop a check in the calling methods (I feel it would be less error prone in the long run).

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:411]: if you drop this check here (in favor of `_dict_to_tuple` - this will be a simpler one liner)

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:606]: it seems to me that this should be enough
```suggestion
function Base.haskey(gd::GroupedDataFrame, key::AbstractDict{<:Union{Symbol, AbstractString}})
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:607]: this test will not be needed if you move it inside `_dict_to_tuple`.

File: src/groupeddataframe/groupeddataframe.jl
Comment (pdeffebach) [reply src/groupeddataframe/groupeddataframe.jl:398]: It was longer before for some reason. Made in line now.

File: src/groupeddataframe/groupeddataframe.jl
Comment (pdeffebach) [reply src/groupeddataframe/groupeddataframe.jl:607]: Yes. all gone noe.

**Comment by pdeffebach** [comment]
Thanks for the review! Ready for another look.

**Comment by bkamins** [comment]
CI fails on 1.0 and on nightly - can you please have a look what causes this (some `convert` errors, but you know your changes better).

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:333]: This should be enough I think as `keys` will produce names as `Symbol` already (maybe this will resolve the CI errors?)
```suggestion
Base.Dict(key::GroupKey) = Dict(pairs(key))
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:614]: also please add it can be `GroupKey` as in the implementation you allow it (and I think it is OK to allow it - just please make sure we test also this case)

**Comment by bkamins** [comment]
Also can you please merge this branch with upstream master (it should be easier than rebasing :smile:) and add an entry to NEWS.md?

**Comment by bkamins** [comment]
CI still fails - can you please check the PR on Julia 1.0 and on Julia nightly to check what causes a problem?

Thank you!

**Comment by pdeffebach** [comment]
Works now!

**Comment by bkamins** [comment]
OK - so can you please just merge it with master and add NEWS.md entry please?

@nalimilan - if you do not comment on this I am going to merge this PR after @pdeffebach adds NEWS.md entry and then work on implementing the conclusions from https://github.com/JuliaData/DataFrames.jl/issues/2305 including handling `Dict`.

**Review by bkamins — APPROVED** [review]

**Comment by bkamins** [comment]
bump :smile: (I would finish it not to have constantly merge it with master)

**Comment by pdeffebach** [comment]
Thanks for pinging me on this! Slipped through. Thanks for the feedback Milan.

**Comment by bkamins** [comment]
also merge conflicts need resolving

**Review by bkamins — APPROVED** [review]

**Review by nalimilan — APPROVED** [review]

**Comment by pdeffebach** [comment]
Conflict was just `News.md`. This can be merged.

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**
